compiler/noirc_evaluator/src/ssa/opt/mem2reg.rs

```diff
@@ -116,13 +116,36 @@ struct PerFunctionContext<'f> {
 
     /// Track a value's last load across all blocks.
     /// If a value is not used in anymore loads we can remove the last store to that value.
-    last_loads: HashMap<ValueId, (InstructionId, BasicBlockId)>,
+    last_loads: HashMap<ValueId, (InstructionId, BasicBlockId, u32)>,
+
+    /// Track whether a load result was used across all blocks.
+    load_results: HashMap<ValueId, PerFuncLoadResultContext>,
+
+    /// Track whether a reference was passed into another entry point
+    /// This is needed to determine whether we can remove a store.
+    calls_reference_input: HashSet<ValueId>,
 
     /// Track whether the last instruction is an inc_rc/dec_rc instruction.
     /// If it is we should not remove any known store values.
     inside_rc_reload: bool,
 }
 
+#[derive(Debug, Clone)]
+struct PerFuncLoadResultContext {
+    /// Reference counter that keeps track of how many times a load was used in other instructions
+    result_counter: u32,
```

Comment:
```suggestion
    uses: u32,
```
Seems to make more sense with the comment above

Reply:
I did this locally as to take advantage of rename symbol.